README.md
# yape

Yet Another pButtons Extractor

Use at your own risk and if you find any issues, please feel free to file it [here](https://github.com/murrayo/yape/issues).

There is a link to how-to videos on [InterSystems Developer Community](https://community.intersystems.com/post/unpacking-pbuttons-yape-update-notes-and-quick-guides).

yape automatically creates charts that can be used for trouble-shooting performance issues or informing performance questions.

Below is the example chart of a weeks data.

![](images/mgstat.Glorefs.png)

![](images/vmstat.Total%20CPU.png)

For more fancy charts or to combine mgstat, vmstat an iostat use [pretty_pButtons](https://github.com/murrayo/MUT)

![](images/temp_Glorefs_and_vmstat_-_Sample_Wednesday_31_Jul_2019.png)


## Yape Installation: Docker Container (recommended)

### Use image from docker hub

Of course, you need docker installed first :)

[https://hub.docker.com/r/yape/yape/]

The same usage information applies, for example:

```
docker pull yape/yape
docker container run --rm -v "$(pwd)":/data yape/yape --help
```

For example to extract mgstat and vmstat quietly and also output csv files:
```
cd <directory with your pButtons file.html>
docker container run --rm -v "$(pwd)":/data yape/yape --mgstat --vmstat -qc /data/<name of your pButtons file.html>
```


### Create image yourself

To avoid any fighting with python versions there is a Dockerfile for building a container

```
cd <some place you want the files>
git clone https://github.com/murrayo/yape.git
cd yape
docker build -t yape2 .
```

You can run the container like this example to get parameters:
```
docker container run --rm -v "$(pwd)":/data yape2 --help
```

For example to extract mgstat and vmstat quietly and also output csv files:
```
cd <directory with your pButtons file.html>
docker container run --rm -v "$(pwd)":/data yape2 --mgstat --vmstat -qc /data/<name of your pButtons file.html>
```

>Note:
>This installs the command line version only, for the interactive version you will need to install locally in your OS (see below).

If you want to remove the container version. Stop all containers and remove images.

## Installation: Local

Requires git, pip and Python installed aready as per Python Version as below. Includes Alpha of the interactive version.

### About Python Version

Python can be a beast about versions, for the simplest experience a container is a good way to go. You have been warned.

See: https://www.anaconda.com/distribution

The following was used to test and build yape for the container version:

```
$ python --version
Python 3.7.4
```

To use pip to install packages to Anaconda Environment
```
$ sudo conda install pip
```

```
$ pip --version
pip 19.2.3 from /Users/mo/opt/anaconda3/lib/python3.7/site-packages/pip (python 3.7)
```

### Install for running locally

```
cd <some place you want the files>
git clone https://github.com/murrayo/yape.git
sudo pip install . --upgrade

yape --help
```

## Other things you can try

### Personal default options

You can specify a configuration file with the
```
yape --config config.example.yml ...
```
parameter. This allows to set things like the default output graph size. See `config.example.yml` for possible options and their default value.
Yape will always check if there is a `~/.yape.yml` file and load that.

### Example: Dotgraphs plotting
To switch to dot graphs instead of line graphs set the style to "". For example, define the following in your config:
```
plotting:
  dim: 16,6
  style: ""
  markersize: 2
```

dim is dimensions in inches.

Choices for style are:
'-' solid line
'--' dashed line
'-.' dash-dotted line
':' dotted line

The defaults for configuration are:
```
plotting:
  dim: 16,6
  style: "-"
  markersize: 1
```

### Weekly overview graphs

To create a week overview graph you can currently parse a number of pbuttons into a file and then plot that:
```
yape --filedb data.db pbuttons1.html
yape --filedb data.db pbuttons2.html
yape --filedb data.db pbuttons3.html
yape --filedb data.db pbuttons4.html
yape --filedb data.db pbuttons5.html
yape --filedb data.db pbuttons6.html
yape --filedb data.db pbuttons7.html
....
yape --filedb data.db --timeframe "2018-06-11 00:00:00,2018-06-17 22:00:00" -a --skip-parse -o testdata/ 
```
Note: the last positional argument (pButtons file name) is not entered when --skip-parse and --filedb are used.

## Experimental interactive version

Change to the base of your checked out yape directory and run:
```
bokeh serve --show yapesrv --args /path/to/your.html
```

This will give you (maybe) an interactive display of the pbuttons passed in. If you run into any errors, feel free to create an issue here: https://github.com/murrayo/yape/issues

## Testing

To run the unittests, use
```
python setup.py nosetests
```
(you might need to use python3 as command if you have multiple versions of python installed)

It will run tests with all pbutton files found in the test-data directory.

## Related Discussion

See the detailed description and discussion [in this article](https://community.intersystems.com/post/yape-yet-another-pbuttons-extractor-and-automatically-create-charts).
Have a look at [InterSystems Developer Community](community.intersystems.com) to learn about InterSystems technology, sharing solutions and staying up-to-date on the latest developments.


CHANGELOG.md
# Changelog


## (unreleased)

### Fix

* Commas in PERFMON names. [murrayo]

### Other

* After blkack. [murrayo]

* Update version and readme. [murrayo]


## 2.2.9 (2020-10-13)

### Other

* Version fix. [murrayo]

* Update setup. [murrayo]

* Doc: new version. [murrayo]

* After black. [murrayo]

* Fix sar and euopean time. [murrayo]

* Doc: Bump version. [murrayo]


## 2.2.8 (2019-08-18)

### Fix

* Wring github repository. [murrayo]

### Other

* Doc: Fix repository. [murrayo]

* Doc: update version. [murrayo]


## 2.2.7 (2019-08-18)

### Changes

* If timeframe, test for max only on timeframe for y axis scale. [murrayo]

* Charts with % y axis top is 100. [murrayo]

### Other

* Doc: update changelog. [murrayo]

* Doc: Commit black changes. [murrayo]

* Merge pull request #20 from casep/master. [murrayo]

  Merging for lighter container image

* Updated to minimise image size. [Carlos "casep" Sepulveda]

* Dockerfile updated for alpine. [Carlos "casep" Sepulveda]

* Relaxed restrictions on versions. [Carlos "casep" Sepulveda]

* Doc: Update readme. [murrayo]

* Doc: Update docs to include images. [murrayo]

* Doc: Update read me. [murrayo]

* Doc: Version change. [murrayo]


## 2.2.6 (2019-07-24)

### Fix

* Some minor fixes after update python libraries. [murrayo]

### Other

* Doc: Update change log. [murrayo]

* Doc: update version. [murrayo]


## 2.2.5 (2019-07-24)

### Changes

* Make date string consistent for windows title (drop decimal places), add short day to title. [murrayo]

* Increase decimal precision for y axis less than 10, eg for windows sec/read is seconds not ms like linux. [murrayo]

### Fix

* Order of case prevents action. [murrayo]

### Other

* Doc: Prepare for docker version update. [murrayo]

* Doc: Bump version. [murrayo]


## 2.2.4 (2019-06-20)

### Changes

* Update line style choices in config file. [murrayo]

### Fix

* Reinstate config file, make some more changes to smarter x and y axis. [murrayo]

* Solve for yyyy dates and yy dates or bail out. [murrayo]

### Other

* Doc: Update change log. [murrayo]

* Doc: Update readme. [murrayo]

* Doc: bump version. [murrayo]


## 2.2.3 (2019-06-03)

### Changes

* Positional arg pButtons html or zip file name NOT required if passing filedb and skip-parse. [murrayo]

### Other

* Doc: Update change log ahead of version bump. [murrayo]

* Default output directory if using previously created database (--skip-parse) [murrayo]

* Doc: Update readme with parameter details. [murrayo]

* Doc: bump version. [murrayo]


## 2.2.2 (2019-05-17)

### Changes

* Some tidy up on the date format for multi-day charting. [murrayo]

* Added device name to chart title for iostat and sar chg: bit more tidy up on date/time handling for charts. [murrayo]

* Updated date handling. [murrayo]

### Other

* Doc: black latest changes. [murrayo]

* [fix] Fixed problem with sar on red hat not charting properly. [murrayo]

* Doc: Bump Version. [murrayo]


## 2.2.1 (2019-05-06)

### Changes

* Some massaging of the chart output to make more consistent and readable. Update AIX formats. Black code. [murrayo]

### Other

* Doc: Updated changelog. [murrayo]

* Quite a lot of updates. - Some layout changes to the charts to make them a bit nicer for reports etc, including adding dates to the heading and some smarter x and y axis - vmstat now produces a Total CPU chart so you don't need to stand on your head and read the id chart. - Added some logic for AIX based on reports of failures. To be honest, I wouldnt be surprised if this caused some other AIX to fail... if so, sorry, please let me know. [murrayo]

* Updated format y axis of vmstat (us, sy, wa) to 100 top, and numbers greater than 1,000. [murrayo]

* Update readme for anaconda. [murrayo]

* Update so that bokeh works with anaconda. [murrayo]

* Merge pull request #19 from murrayo/Adding-styles-to-charts. [murrayo]

  Adding styles to charts

* Merge branch 'master' of https://github.com/murrayo/yape into Adding-styles-to-charts. [murrayo]

  # Conflicts:
  #	yape/plotpbuttons.py

* Merge branch 'master' of github.com:murrayo/yape. [Fabian]

* Merge pull request #15 from kazamatzuri/master. [Fabian]

  doc: add github hint

* Doc: add github hint. [Fabian]

* Fix version indication for black. [Fabian]

* Doc: fix formatting in contributing. [Fabian]

* Doc: versioning adjustment, no auto commit. [Fabian]

* Make charts dot chart to see effects, others may follow. [murrayo]


## 2.2.0 (2019-02-27)

### Changes

* Hooked up black with pre-commit semantic versioning contributing document. [Fabian]

### Fix

* Version numbering config. [Fabian]

### Other

* Bump version: 2.1.0 → 2.2.0. [Fabian]

* Bump version: 2.0.8 → 2.1.0. [Fabian]

* Going black. [Fabian]


## 2.0.8 (2019-02-20)

### Other

* Fix sar-u plotting on hp-ux; implement sar-u plotting. [Fabian]

* Pep8 compliance. [Fabian]


## 2.0.7 (2019-02-06)

### Other

* Prevent scientific notation. [Fabian]


## 2.0.5 (2018-12-31)

### Other

* Syntax errors ... [Fabian]

* Declarative pipeline update. [Fabian]


## 2.0.4 (2018-12-31)

### Other

* Added jenkins agent label. [Fabian]


## 2.0.3 (2018-12-31)

### Other

* Pipeline, jenkins docker file. [Fabian]

* Jenkinsfile, updated readme. [Fabian]

* Remove dind. [Fabian]

* Try just building. [Fabian]

* Names.. [Fabian]

* Gitlab pipeline 1st attempt. [Fabian]


## 2.0.2 (2018-11-30)

### Other

* Improve error message when parsing bad mgstat data. [Fabian]

* Vscode configs. [Fabian]

* Add entrypoint for debugging. [Fabian]

* Updated readme, fixed reference to docker image and link. [Fabian]

* Fix old version lookup. [Fabian]

* Updated readme. [Fabian]

* Docker image build updates. [Fabian]


## 2.0.0 (2018-10-09)

### Other

* Versioning stuff. [Fabian]

* Eporting parse_args. [Fabian]

* Updated readme. [Fabian]

* Making dot graphs configurable in the config file. [Fabian]

* Fix conditional to make sure minor time ticks are rendered. [Fabian]


yape/main.py
# os methods for manipulating paths
import os
import argparse

import sys
import csv

import sqlite3

import logging
import tempfile
import zipfile
import yaml
from pathlib import Path

from yape.parsepbuttons import parsepbuttons
from yape.plotpbuttons import mgstat, vmstat, iostat, perfmon, sard, monitor_disk, saru
from pkg_resources import get_distribution, DistributionNotFound


def getVersion():
    v = ""
    try:
        v = get_distribution("yape").version
    except DistributionNotFound:
        v = ""
    pass
    return v


def read_config(cfgfile, config):
    if cfgfile is None:
        cfgfile = os.path.join(Path.home(), ".yape.yml")
    if os.path.isfile(cfgfile):
        with open(cfgfile, "r") as ymlfile:
            cfg = yaml.load(ymlfile, Loader=yaml.FullLoader)
            logging.debug(cfg)
            return {**config, **cfg}
    return config


def fileout(db, config, section):
    fileprefix = config["fileprefix"]
    basefilename = config["basefilename"]

    c = db.cursor()
    c.execute("SELECT name FROM sqlite_master WHERE type='table' AND name=?", [section])
    if len(c.fetchall()) == 0:
        return None
    file = os.path.join(basefilename, fileprefix + section + ".csv")
    print("exporting " + section + " to " + file)
    c.execute('select * from "' + section + '"')
    columns = [i[0] for i in c.description]

    with open(file, "w") as f:
        csvWriter = csv.writer(f)
        csvWriter.writerow(columns)
        csvWriter.writerows(c)


def ensure_dir(file_path):
    directory = os.path.dirname(file_path)
    if not os.path.exists(directory):
        os.makedirs(directory)


def fileout_splitcols(db, config, section, split_on):
    fileprefix = config["fileprefix"]
    basefilename = config["basefilename"]
    c = db.cursor()
    c.execute("SELECT name FROM sqlite_master WHERE type='table' AND name=?", [section])
    if len(c.fetchall()) == 0:
        return None
    c.execute("select distinct " + split_on + ' from "' + section + '"')
    rows = c.fetchall()
    for column in rows:
        c.execute(
            'select * from "' + section + '" where ' + split_on + "=?", [column[0]]
        )
        file = os.path.join(
            basefilename, fileprefix + section + "." + column[0] + ".csv"
        )
        print("exporting " + section + "-" + column[0] + " to " + file)
        columns = [i[0] for i in c.description]
        with open(file, "w") as f:
            csvWriter = csv.writer(f)
            csvWriter.writerow(columns)
            csvWriter.writerows(c)


def parse_args(args):
    parser = argparse.ArgumentParser(description="Yape")
    parser.add_argument(
        "-v",
        "--version",
        dest="version",
        help="display version information",
        action="version",
        version="%(prog)s " + getVersion(),
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        "pButtons_file_name", nargs="?", help="path to pButtons file to use"
    )
    group.add_argument(
        "--skip-parse",
        dest="skipparse",
        help="disable parsing; requires filedb to be specified to supply data",
        action="store_true",
    )
    parser.add_argument(
        "--filedb",
        help="use specific file as DB, useful to be able to used afterwards or as standalone datasource.",
    )
    parser.add_argument(
        "-c",
        dest="csv",
        help="will output the parsed tables as csv files. useful for further processing. will currently create: mgstat, vmstat, sar-u. sar-d and iostat will be output per device",
        action="store_true",
    )
    parser.add_argument(
        "--mgstat", dest="graphmgstat", help="plot mgstat data", action="store_true"
    )
    parser.add_argument(
        "--vmstat", dest="graphvmstat", help="plot vmstat data", action="store_true"
    )
    parser.add_argument(
        "--iostat", dest="graphiostat", help="plot iostat data", action="store_true"
    )
    parser.add_argument(
        "--sard", dest="graphsard", help="plot sar-d data", action="store_true"
    )
    parser.add_argument(
        "--saru", dest="graphsaru", help="plot sar-u data", action="store_true"
    )
    parser.add_argument(
        "--monitor_disk",
        dest="monitor_disk",
        help="plot disk data from monitor (vms)",
        action="store_true",
    )
    parser.add_argument(
        "--perfmon", dest="graphperfmon", help="plot perfmon data", action="store_true"
    )
    parser.add_argument(
        "--timeframe",
        dest="timeframe",
        help='specify a timeframe for the plots, i.e. --timeframe "2018-05-16 00:01:16,2018-05-16 17:04:15"',
    )
    parser.add_argument(
        "--prefix",
        dest="prefix",
        help="specify output file prefix (this is for the filename itself, to specify a directory, use -o)",
    )
    parser.add_argument(
        "--plotDisks", dest="plotDisks", help="restrict list of disks to plot"
    )
    parser.add_argument(
        "--log",
        dest="loglevel",
        help="set log level:DEBUG,INFO,WARNING,ERROR,CRITICAL. The default is INFO",
    )
    parser.add_argument(
        "-a", "--all", dest="all", help="graph everything", action="store_true"
    )
    parser.add_argument(
        "-q", "--quiet", dest="quiet", help="no stdout output", action="store_true"
    )
    parser.add_argument(
        "-o",
        "--out",
        dest="out",
        help="specify base output directory, defaulting to <pbuttons_name>/",
    )
    parser.add_argument(
        "--config",
        dest="configfile",
        help="specify the location of a config file. ~/.yape.yml is used by default.",
    )
    return parser.parse_args(args)


def yape2(args=None):
    if args == None:
        args = parse_args(sys.argv[1:])
    try:
        if args.loglevel is not None:
            loglevel = getattr(logging, args.loglevel.upper(), None)
            if not isinstance(loglevel, int):
                raise ValueError("Invalid log level: %s" % args.loglevel)
            logging.basicConfig(level=loglevel)
        else:
            logging.basicConfig(level=getattr(logging, "INFO", None))
        if args.quiet:
            logger = logging.getLogger()
            logger.disabled = True
        if args.skipparse:
            if args.filedb is None:
                logging.error("filedb required with skip-parse set")
                return -1
            if args.out is None:
                args.out = "output"  # Default output directory when no filename
        if args.filedb is not None:
            db = sqlite3.connect(args.filedb)
        else:
            db = sqlite3.connect(":memory:")
            db.execute("pragma journal_mode=wal")
            db.execute("pragma synchronous=0")

        if args.prefix is not None:
            fileprefix = args.prefix
        else:
            fileprefix = ""

        if not args.skipparse:
            # it's unrealistic to assume we wil have enough memory to hold the extracted pbuttons file
            # so we extract it to a temp directory and extract it there
            if not args.pButtons_file_name:
                logging.error(
                    "pbuttons html file name required unless skip-parse and filedb filename supplied"
                )
                exit(1)

            if args.pButtons_file_name.split(".")[-1] == "zip":

                with tempfile.TemporaryDirectory() as tmpdir:
                    dest = tmpdir + os.sep
                    ensure_dir(tmpdir + os.sep)
                    with open(args.pButtons_file_name, "rb") as f:
                        zf = zipfile.ZipFile(f)
                        zf.extractall(dest)
                        logging.debug("using temp directory:" + dest)
                        temppbfile = None
                        for root, dirs, files in os.walk(dest):
                            for file_ in files:
                                # this will only grab the first html file in the zipfile
                                # we assume for now people are not evil in what they pass in ....
                                if "html" in file_:
                                    temppbfile = os.path.join(root, file_)
                                    continue
                        if temppbfile is not None:
                            parsepbuttons(temppbfile, db)
                        else:
                            logging.error(
                                "passed in zipfile with no included pbuttons html file"
                            )
                            exit(1)
            parsepbuttons(args.pButtons_file_name, db)

        if args.out is not None:
            basefilename = args.out
        else:
            basefilename = args.pButtons_file_name.split(".")[0]

        if args.plotDisks is not None:
            plotDisks = args.plotDisks
        else:
            plotDisks = ""

        # a place to hold global configurations/settings
        # makes it easier to extend functionality to carry
        # command line parameters to subfunctions...
        config = {}
        config["quiet"] = args.quiet

        # doing config file read here, because we want the quiet flag to be overwritable in the config
        # but not the below config settings
        config = read_config(args.configfile, config)
        logging.debug(config)
        config["fileprefix"] = fileprefix
        config["plotDisks"] = plotDisks
        config["timeframe"] = args.timeframe
        config["basefilename"] = basefilename

        if args.csv:
            ensure_dir(basefilename + os.sep)
            fileout(db, config, "mgstat")
            fileout(db, config, "vmstat")
            fileout_splitcols(db, config, "iostat", "Device")
            fileout_splitcols(db, config, "sar-d", "DEV")
            fileout(db, config, "perfmon")
            fileout(db, config, "sar-u")

        # plotting
        if args.graphsard or args.all:
            ensure_dir(basefilename + os.sep)
            sard(db, config)

        if args.graphsaru or args.all:
            ensure_dir(basefilename + os.sep)
            saru(db, config)

        if args.graphmgstat or args.all:
            ensure_dir(basefilename + os.sep)
            mgstat(db, config)

        if args.graphvmstat or args.all:
            ensure_dir(basefilename + os.sep)
            vmstat(db, config)

        if args.monitor_disk or args.all:
            ensure_dir(basefilename + os.sep)
            monitor_disk(db, config)

        if args.graphiostat or args.all:
            ensure_dir(basefilename + os.sep)
            iostat(db, config)

        if args.graphperfmon or args.all:
            ensure_dir(basefilename + os.sep)
            perfmon(db, config)

    except OSError as e:
        print("Could not process pButtons file because: {}".format(str(e)))


if __name__ == "__main__":
    # execute only if run as the entry point into the program
    yape2()
